boot/makebootable.go

```diff
@@ -314,11 +317,21 @@ func MakeRecoverySystemBootable(model *asserts.Model, rootdir string, relativeRe
 		for k, v := range recoveryCmdlineArgs {
 			recoveryBlVars[k] = v
 		}
+
+		trustedAssets, err := tbl.TrustedAssets()
+		if err == nil && len(trustedAssets) > 0 {
+			seedShimPath = trustedAssets[0]
+		}
 	}
 
 	if err := rbl.SetRecoverySystemEnv(relativeRecoverySystemDir, recoveryBlVars); err != nil {
 		return fmt.Errorf("cannot set recovery system environment: %v", err)
 	}
+
+	if seedShimPath != "" {
+		SetEfiBootVariables(seedShimPath)
+	}
```

Comment:
Should `RecoveryBootChain(kernelPath)` be used instead of `TrustedAssets()`? The former requires a `kernelPath`, as it adds on a new `"kernel.efi"` boot file at the end of the chain, but we don't care about this as we really only want the first entry from `getGrubRecoveryModeTrustedAssets()`.

More simply, we want the `shimFallbackBinary`, if it exists, else `shimBinary`. Previous commits (From xnox) in this PR add support for shim fallback in grub (by modifying `getGrubRecoveryModeTrustedAssets`) in exactly the way we want.

We could add a dedicated (exported) function for either the `Bootloader` interface or the `RecoveryAwareBootloader` (as opposed to `TrustedAssetsBootloader`) interface. Setting EFI boot variables is orthogonal to secure boot and trusted assets, so this could be a good distinction to make.

`RecoveryBootChain()` is fairly clearly what we are looking for, aside from requiring a kernel path and outputting `[]BootFile` instead of `[]string`, but this is a method on the `TrustedAssetsBootloader` interface as well, which suggests that perhaps recovery booting is tied to trusted assets anyway, and we should just continue to use `TrustedAssets()`, which provides exactly what we need.

Reply:
I feel like a function that returns an array in the first place is probably the wrong thing.

I would add a function to `bootloader/grub.go` that returns an `efi.LoadOption` directly. It knows about EFI, so we should be able to build it.

Or return an abstract type that can build an  `efi.LoadOption`.

That way we can make the bootloader configure things more complicated, like passing the path to the bootloader as parameter.

Reply:
But we could also merge with a big FIXME.

Reply:
(also I have in my mind that while secboot usually implies UEFI, UEFI does not always imply secboot. So I can imagine  a boot loader not implementing `TrustedAssetsBootloader`, but still wanting to define an UEFI boot entry.)

Reply:
I'm considering an addition to the `Bootloader interface` (as opposed to `TrustedAssetBootloader`), which I was thinking could return the shim path including the `rootdir`, which common code would then use to do all the EFI variable heavy lifting. But it's true, having it return a serialized `efi.LoadOption` could be nice if we ever want to set optional data, very good point!

Actually, perhaps it would be best to define a new `UefiBootloader` which has the method `EfiLoadOption()` which requires `rootdir` to be set, and returns a `efi.LoadOption` with a device path built from`/run/mnt/ubuntu-seed/EFI/ubuntu/shimx64.efi` (or what is appropriate depending on gadget and architecture).

I don't love that bootloaders have to know about `go-efilib`, but the code to go from shim filepath to load option really isn't much, so perhaps this is the best approach. I'll start out with just grub implementing it, but if there are other bootloaders which should also set EFI boot variables, please let me know!

Reply:
See most recent commit